add WarningManager (#4473)

ricochet opened the pull request:
See #4473

Ran core, sanity, and other test suites. Only the tests where I have not configured certain tools to run failed (test_d8_path, test_crunch, test_llvm_lit).

ricochet commented on emcc.py:
```diff
@@ -460,18 +460,13 @@ def is_valid_abspath(path_name):
-          newargs[i] = ''
-          absolute_warning_shown = True
-        elif newargs[i] in ['-l', '-L', '-I']:
+        if newargs[i] in ['-l', '-L', '-I']:
```
I think this could move down to line 682 and modify newargs in place. I don't know the history here, but as written, I'm looping through newargs thrice. It didn't make sense to move this into WarningManager unless it changed to be CommandLineMunger and warning enablement was added to the Settings object.

dschuff replied:
Not sure, that would probably be a question for @kripken but I think he's got limited availability for a few days.

kripken replied:
That does seem possible, but I think it's ok to leave it as for now. I'm not worried about scanning newargs several times, shouldn't matter in terms of speed.

dschuff commented on tools/shared.py:
```diff
@@ -772,6 +772,32 @@ def clean_temp():
+    assert cmd_args
+
+    for i in range(len(cmd_args)):
+      if cmd_args[i] == '-Wwarn-absolute-paths':
```
I think the 'warn' part of this string is redundant; the more conventional form would be `-Wabsolute-paths` and `-Wno-absolute-paths`.

dschuff replied:
Oh, I just realized that the absolute path flags were pre-existing. So I guess we maybe shouldn't change them. But the new ones can at least match the usual convention.

dschuff commented on tools/shared.py:
```diff
@@ -772,6 +772,32 @@ def clean_temp():
     prepare_to_clean_temp(EMSCRIPTEN_TEMP_DIR) # this global var might change later
   return EMSCRIPTEN_TEMP_DIR
 
+class WarningManager:
```
I think it might make more sense to have `WarningManager` just implement a `warn()` method. Over time we could convert all the emscripten-specific warnings to use it.
There would be 2 advantages to that. First, you wouldn't have to add yet-another case to the if conditions that emit the warning. But even better, you could then support flags like `-W` (disable all warnings) and other centralized ways of controlling warnings.
You'd sometimes still have to pass it the type of warning in some cases, so it could be something like this:

```
class WarningManager:
  def warn(type, text):
    if <type is not suppressed>:
      logging.warning(text)
```
and in the emcc code,
```
shared.WarningMananger.warn(shared.WarningManager.WARN_SEPARATE_ASM, 'warning!')
```
Or maybe even get rid of the enums and just use a string that matches the warning name in the flag, like 'absolute-paths'. That might make the `warn` code cleaner (see my suggestion below).
Not sure which of those is better though, I still generally like enums more than strings. Hmm...

ricochet replied:
I agree, having a warn method is a better implementation. I'll work on that and see if using a string dict can help clean up the code.

dschuff commented on tools/shared.py:
```diff
@@ -772,6 +772,32 @@ def clean_temp():
+      elif cmd_args[i] == '-Wno-warn-absolute-paths':
+        cmd_args[i] = ''
+        WarningManager.WARN_ABSOLUTE_PATHS = False
+      elif cmd_args[i] == '-Wno-warn-separate-asm':
```
We should support both the `yes` and `no` forms of each of the flags, as you did with `absolute-paths`. Rather than explicit cases for every possible flag, it might make more sense to just see if the flag starts with "-W" and then see if it next has "no-" and then switch on the rest of the string. Actually you wouldn't even need to do  a switch or cases. You could just store the string in a dict with a boolean saying whether it was enabled or disabled and look that up again when `warn` is called. We would still want a dict containing all of the known flags, so that any flags with unknown warning options can be passed through to the underlying clang.

dschuff commented on tests/test_other.py:
```diff
@@ -5755,6 +5755,13 @@ def test_no_warn_exported_jslibfunc(self):
     out, err = Popen([PYTHON, EMCC, path_from_root('tests', 'hello_world.c'), '-s', 'DEFAULT_LIBRARY_FUNCS_TO_INCLUDE=["alGetError"]', '-s', 'EXPORTED_FUNCTIONS=["_main", "_alGetError"]'], stdout=PIPE, stderr=PIPE).communicate()
     self.assertNotContained('''function requested to be exported, but not implemented: "_alGetError"''', err)
 
+  def test_suppress_almost_asm_warning(self):
```
We should also have one test that passes both the `-Wfoo` and `-Wno-foo` forms of one of the flags to check that the second one overrides the first.

ricochet replied:
Good catch. I fixed this in my last commit.

kripken commented on emcc.py:
```diff
@@ -460,18 +460,13 @@ def is_valid_abspath(path_name):
+      # Scan and strip emscripten specific cmdline warning flags
+      # This needs to run before other cmdline flags have been parsed, so that warnings are properly printed during arg parse
+      newargs = shared.WarningManager.capture_warnings(newargs)
+      absolute_warning_shown = False
```
could the warning manager perhaps also track if the warning was shown? if it handled printing the warnings itself, then it could handle showing each at most once.

ricochet commented:
Thanks for the suggestions. They were all major improvements. I think I've addressed everything with my last commit.

dschuff commented on emcc.py:
```diff
@@ -955,7 +948,7 @@ def check(input_file):
 
       if separate_asm and final_suffix != 'html':
-        logging.warning("--separate-asm works best when compiling to HTML. otherwise, you must yourself load the '.asm.js' file that is emitted separately, and must do so before loading the main '.js` file")
+        shared.WarningManager.warn(shared.WarningType.SEPARATE_ASM, "--separate-asm works best when compiling to HTML. Otherwise, you must yourself load the '.asm.js' file that is emitted separately, and must do so before loading the main '.js' file. Pass in '-Wno-warn-separate-asm' to disable this warning.")
```
To fit in even more with clang, we can use clang's format for displaying the warning flag, e.g.
` warning: --separate-asm works best ... '.js' file [-Wwarn-separate-asm]`. You could even put that inside of `WarningManager.warn` so you wouldn't have to repeat it here.

dschuff commented on tools/shared.py:
```diff
@@ -772,6 +772,54 @@ def clean_temp():
+    if key in self:
+      return key
+
+WarningType = Enum(['ABSOLUTE_PATHS', 'SEPARATE_ASM', 'ALMOST_ASM'])
```
Come to think of it, it seems a little funny that we are enumerating all of the warning types (and constructing their corresponding flag names) here, but we are are defining the text elsewhere. It seems like it should be one of the following cases:
1. there is no central list of all the warnings (in which case both the warning/flag name and warning text are passed to `warn()`). Then `WarningManager` just becomes a utility class that formats the message and tracks whether warnings are enabled and printed, and handles flags based on names/enums. The external code defines which warnings exist, their names, and the conditions under which they fire.
2. WarningManager has the list of all warnings, along with their enums/names and warning text. Then the external code only determines the conditions under which they fire.

2 would seem to be better if there is more than one condition or external location that can cause a warning (it would avoid duplication), but that doesn't seem to be the case currently. Do either of you have any intuition about whether that will be the case in the future?

ricochet replied:
I liked the idea of predefining warning types if only to make it easier to find documentation within the code on what warnings exist. De-optimizing to ASM_JS=2 (ALMOST_ASM) appears in two different places, but it might be that each would prefer to add their own customized message, e.g. "ALLOW_MEMORY_GROWTH caused this warning to display because its not supported in the asm.js spec". The WarningManager could supply a default message and warn will use the default message if parameter message=None. 

For option 1, I would need to add another API on WarningManager to disable_warning (and also enabled_warning for completeness) to handle some warnings being disabled by default (absolute paths). Default warning status will need to be initialized on the WarningManager prior to emcc calling capture_warnings. It seems cleaner for the WarningManager to manage the default status of all warnings vs having a block at the top of emcc before cmd arg parsing.

dschuff replied:
I like that idea; the warning manager has default warning text, which can be overridden (or appended to?) by the caller.

ricochet commented:
Sorry for disappearing. Hopefully my last commit addressed the remaining issues. 
- default message
- clang's format for displaying the warning flag
- swapped out the enum for a dict.

kripken commented on tools/shared.py:
```diff
@@ -772,6 +772,58 @@ def clean_temp():
+
+  @staticmethod
+  def capture_warnings(cmd_args):
+    assert cmd_args
```
i don't think we need this, it will throw later down if that is the case anyhow

kripken commented:
Thanks, this looks good to me. Just one tiny comment above. Otherwise, please verify this passes `tests/runner.py other` (can ignore stuff like crunch tests if you don't have that installed, but everything else, in particular the `test_emcc*` set of tests, should pass).

ricochet commented:
Deleted the assert and `python tests/runner.py other` ran with one failure. It does not appear to be related. 

```
======================================================================
FAIL: test_llvm_lit (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/bahaye/unicorn_tools/emsdk/osx/emsdk_portable_incoming/emscripten/incoming/tests/test_other.py", line 3359, in test_llvm_lit
    assert p.returncode == 0, 'LLVM tests must pass with exit code 0'
AssertionError: LLVM tests must pass with exit code 0

----------------------------------------------------------------------
Ran 241 tests in 1849.752s

FAILED (failures=1)
```

kripken commented:
Thanks, and yeah, that error is unrelated (I think that test requires llvm to be built with tests, but anyhow no possible connection to the changes here).

lgtm. @dschuff, you had some comments before, did you want to take another look? Otherwise we can merge this.

cancerberoSgx commented:
I think there are still warnings missing from other scripts than emcc.py, this is annoying me lately (https://github.com/emscripten-core/emscripten/blob/6d6cb961cf3264747bd3a4ead95fc631f7f0de66/emconfigure.py#L36)

I wonder how hard would be to use the manager in those files. No much python experience but will give it a try when I have time. thanks

dschuff commented:
It looks like emconfigure.py is already importing things from `shared`, so in principle I don't think it would be a problem for it to also import and use `shared.WarningManager`.
Looking at that specific output though, it seems odd. Usually such usage messages are only used when the arguments are wrong and then the whole program just exits. Here it looks like it just prints anytime you use something other than `cmake` or `configure` but it explicitly says you can use any command. I think my preference would be just not to print anything when the tool is used correctly; and otherwise print a message and exit.

sbc100 commented:
out of interest how are you using `emconfigure`?   Are you using it in a non-cmake, non-autotools context?

cancerberoSgx commented:
> Usually such usage messages are only used when the arguments are wrong and then the whole program just exits.


> out of interest how are you using emconfigure? Are you using it in a non-cmake, non-autotools context?

No, I'm using it always followed by ./configure or cmake commands
After little research:

 * if the command is different than ./configure or cmake it always prints no matter if errors:

 docker run --rm --workdir /code/ -v "$PWD/:/code" "trzeci/emscripten:latest" emconfigure  ls

 docker run --rm --workdir /code/ -v "$PWD/:/code" "trzeci/emscripten:latest" emconfigure  ls dontexists

 * it prints to stderr which in my case that I'm using ansi filters to color errors and warnings in the terminal makes it more visible

 * ( I tried with other commands that which might make sense to use it like autoconf, ./libtool or make) ?
, 
 * the message doesn't depend on if ./configure or cmake throw error but it seems to be caused by commands like the following:
```
  emconfigure ./configure CFLAGS="$CFLAGS"  
```
And I think the problem is with `CFLAGS="$CFLAGS" `.Looking at one of the libraries config.log files it says
```
This file contains any messages produced by compilers while
running configure, to aid debugging if configure makes a mistake.

It was created by lcms2 configure 2.9, which was
generated by GNU Autoconf 2.69.  Invocation command line was

  $ ./configure --prefix=/code/emscripten_prefix --disable-shared --enable-static --without-threads --disable-openmp CFLAGS=-I/code/emscripten_prefix/include -I/code/emscripten_prefix/include/freetype2 -I/code/emscripten_prefix/include/openjpeg-2.3 -I/code/emscripten_prefix/include/libraw -I/code/emscripten_prefix/include/libde265 -s BINARYEN_TRAP_MODE=clamp -s ALLOW_MEMORY_GROWTH=1 -Wno-almost-asm -s ERROR_ON_UNDEFINED_SYMBOLS=1 -Werror=implicit-function-declaration  -O3 --llvm-lto 3 --llvm-opts 3 -fno-rtti -fno-exceptions -s DISABLE_EXCEPTION_CATCHING=1 -s NODEJS_CATCH_EXIT=0 -s ASSERTIONS=0 -s AGGRESSIVE_VARIABLE_ELIMINATION=1 PKG_CONFIG_PATH=/code/emscripten_prefix/lib/pkgconfig
```

Notice that the quotes were removed in CFLAG="" although the configure script runs successfully I think probably such a large command is confusing emconfiure. Anyway I will simplify the commands since has lots of unneeded args and play with the quotes to silence it but in any case it would be perfect to have a setting.

Jus in case the scripts are available here https://github.com/cancerberoSgx/magica/blob/master/magick-wasm/build/emscripten-scripts/build-lcms.sh

Probably I have something wrong though although the builds are success-full. emcc is working better and better each day :)

sbc100 commented:
emconfigure should only print that message if "configure" or "cmake" do not appear in argv[1].

if you are passing "./configure" as the first arg and still seeing this message something is going wrong.  Can you edit the script and add `print(sys.argv[1])` right before it prints that message
